File: app/src/main/java/com/baidu/tv/player/auth/LoginActivity.java
package com.baidu.tv.player.auth;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;

import com.baidu.tv.player.R;
import com.baidu.tv.player.model.DeviceCodeResponse;
import com.baidu.tv.player.ui.main.MainActivity;
import com.baidu.tv.player.utils.QRCodeUtils;

/**
 * 登录界面
 */
public class LoginActivity extends FragmentActivity {
    
    private AuthViewModel viewModel;
    private ImageView ivQrCode;
    private ProgressBar pbLoading;
    private LinearLayout llError;
    private Button btnRetry;
    private TextView tvStatus;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        
        initViews();
        initViewModel();
    }
    
    private void initViews() {
        ivQrCode = findViewById(R.id.iv_qr_code);
        pbLoading = findViewById(R.id.pb_loading);
        llError = findViewById(R.id.ll_error);
        btnRetry = findViewById(R.id.btn_retry);
        tvStatus = findViewById(R.id.tv_status);
        
        btnRetry.setOnClickListener(v -> viewModel.startLogin());
    }
    
    private void initViewModel() {
        viewModel = new ViewModelProvider(this).get(AuthViewModel.class);
        
        viewModel.getAuthState().observe(this, authState -> {
            switch (authState.getStatus()) {
                case LOADING:
                    showLoading();
                    tvStatus.setText(authState.getMessage());
                    break;
                    
                case DEVICE_CODE_RECEIVED:
                    DeviceCodeResponse response = (DeviceCodeResponse) authState.getData();
                    showQRCode(response);
                    // 开始轮询
                    viewModel.startPolling(response.getDeviceCode());
                    tvStatus.setText("请使用手机百度网盘APP扫描二维码登录");
                    break;
                    
                case POLLING:
                    // 保持二维码显示，状态不变
                    break;
                    
                case AUTHENTICATED:
                    Toast.makeText(this, "登录成功", Toast.LENGTH_SHORT).show();
                    startMainActivity();
                    break;
                    
                case UNAUTHENTICATED:
                    // 如果是刷新失败或其他原因导致的未认证，重新开始登录流程
                    viewModel.startLogin();
                    break;
                    
                case ERROR:
                    showError();
                    tvStatus.setText(authState.getMessage());
                    break;
            }
        });
        
        // 开始登录流程
        viewModel.startLogin();
    }
    
    private void showLoading() {
        pbLoading.setVisibility(View.VISIBLE);
        llError.setVisibility(View.GONE);
        ivQrCode.setImageBitmap(null);
    }
    
    private void showQRCode(DeviceCodeResponse response) {
        pbLoading.setVisibility(View.GONE);
        llError.setVisibility(View.GONE);
        
        String url = response.getFullVerificationUrl();
        if (url != null) {
            Bitmap qrCode = QRCodeUtils.createQRCodeBitmap(url, 250, 250);
            ivQrCode.setImageBitmap(qrCode);
        }
    }
    
    private void showError() {
        pbLoading.setVisibility(View.GONE);
        llError.setVisibility(View.VISIBLE);
        ivQrCode.setImageBitmap(null);
    }
    
    private void startMainActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/baidu/tv/player/kt/ui/login/LoginActivity.kt
package com.baidu.tv.player.kt.ui.login

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.baidu.tv.player.kt.R
import com.baidu.tv.player.kt.auth.LoginUiState
import com.baidu.tv.player.kt.auth.LoginViewModel
import com.baidu.tv.player.kt.databinding.ActivityLoginBinding
import com.baidu.tv.player.kt.ui.main.MainActivity
import com.baidu.tv.player.kt.util.QRCodeUtils
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

/**
 * 登录界面：展示设备码二维码，轮询授权状态。
 *
 * 用 ViewBinding + [repeatOnLifecycle] 收集 [LoginViewModel.uiState]（替代 Java 版 LiveData.observe）。
 */
@AndroidEntryPoint
class LoginActivity : FragmentActivity() {

    private val viewModel: LoginViewModel by viewModels()

    private lateinit var binding: ActivityLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.btnRetry.setOnClickListener { viewModel.startLogin() }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state -> render(state) }
            }
        }

        viewModel.startLogin()
    }

    private fun render(state: LoginUiState) = when (state) {
        is LoginUiState.Loading -> {
            showLoading()
            binding.tvStatus.text = state.message
        }
        is LoginUiState.DeviceCodeReceived -> {
            showQrCode(state.response)
            binding.tvStatus.text = getString(R.string.login_scan_prompt)
        }
        LoginUiState.Polling -> Unit // 保持二维码显示
        LoginUiState.Authenticated -> {
            Toast.makeText(this, R.string.login_success, Toast.LENGTH_SHORT).show()
            startMainActivity()
        }
        is LoginUiState.Unauthenticated -> {
            binding.tvStatus.text = state.message
            viewModel.startLogin()
        }
        is LoginUiState.Error -> {
            showError()
            binding.tvStatus.text = state.message
        }
    }

    private fun showLoading() {
        binding.pbLoading.visibility = View.VISIBLE
        binding.llError.visibility = View.GONE
        binding.ivQrCode.visibility = View.GONE
        binding.ivQrCode.setImageDrawable(null)
    }

    private fun showQrCode(response: com.baidu.tv.player.kt.model.DeviceCodeResponse) {
        binding.pbLoading.visibility = View.GONE
        binding.llError.visibility = View.GONE
        response.fullVerificationUrl?.let { url ->
            QRCodeUtils.createQRCodeBitmap(url, QR_CODE_SIZE_PX, QR_CODE_SIZE_PX)?.let { qrCode ->
                binding.ivQrCode.setImageBitmap(qrCode)
                binding.ivQrCode.visibility = View.VISIBLE
            }
        }
    }

    private fun showError() {
        binding.pbLoading.visibility = View.GONE
        binding.llError.visibility = View.VISIBLE
        binding.ivQrCode.visibility = View.GONE
        binding.ivQrCode.setImageDrawable(null)
        // TV 遥控器场景：错误状态必须把焦点交给重试按钮，否则 OK 键无响应
        binding.btnRetry.requestFocus()
    }

    private fun startMainActivity() {
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    private companion object {
        const val QR_CODE_SIZE_PX = 512
    }
}
